packages/schematics/angular/utility/workspace.ts

```diff
@@ -10,60 +10,109 @@ import { json, workspaces } from '@angular-devkit/core';
-      return tree.exists(path);
-    },
-  };
+const DEFAULT_WORKSPACE_PATH = '/angular.json';
+
+// re-export the workspace definition types for convenience
+export type WorkspaceDefinition = workspaces.WorkspaceDefinition;
+export type ProjectDefinition = workspaces.ProjectDefinition;
+export type TargetDefinition = workspaces.TargetDefinition;
+
+/**
+ * A {@link workspaces.WorkspaceHost} backed by a Schematics {@link Tree} instance.
+ */
+class TreeWorkspaceHost implements workspaces.WorkspaceHost {
+  constructor(private readonly tree: Tree) {}
+
+  async readFile(path: string): Promise<string> {
+    return this.tree.readText(path);
+  }
+
+  async writeFile(path: string, data: string): Promise<void> {
+    if (this.tree.exists(path)) {
+      this.tree.overwrite(path, data);
+    } else {
+      this.tree.create(path, data);
+    }
+  }
+
+  async isDirectory(path: string): Promise<boolean> {
+    // approximate a directory check
+    return !this.tree.exists(path) && this.tree.getDir(path).subfiles.length > 0;
+  }
+
+  async isFile(path: string): Promise<boolean> {
+    return this.tree.exists(path);
+  }
 }
 
+/**
+ * Updates the workspace file (`angular.json`) found within the root of the schematic's tree.
+ * The workspace object model can be directly modified within the provided updater function
+ * with changes being written to the workspace file after the updater function returns.
+ * The spacing and overall layout of the file (including comments) will be maintained where
+ * possible when updating the file.
+ *
+ * @param updater An update function that can be used to modify the object model for the
+ * workspace. A {@link WorkspaceDefinition} is provided as the first argument to the function.
+ */
 export function updateWorkspace(
-  updater: (workspace: workspaces.WorkspaceDefinition) => void | Rule | PromiseLike<void | Rule>,
-): Rule;
-export function updateWorkspace(workspace: workspaces.WorkspaceDefinition): Rule;
-export function updateWorkspace(
-  updaterOrWorkspace:
-    | workspaces.WorkspaceDefinition
-    | ((workspace: workspaces.WorkspaceDefinition) => void | Rule | PromiseLike<void | Rule>),
+  updater: (workspace: WorkspaceDefinition) => void | Rule | PromiseLike<void | Rule>,
 ): Rule {
   return async (tree: Tree) => {
-    const host = createHost(tree);
+    const host = new TreeWorkspaceHost(tree);
 
-    if (typeof updaterOrWorkspace === 'function') {
-      const { workspace } = await workspaces.readWorkspace('/', host);
+    const { workspace } = await workspaces.readWorkspace(DEFAULT_WORKSPACE_PATH, host);
 
-      const result = await updaterOrWorkspace(workspace);
+    const result = await updater(workspace);
 
-      await workspaces.writeWorkspace(workspace, host);
+    await workspaces.writeWorkspace(workspace, host);
 
-      return result || noop;
-    } else {
-      await workspaces.writeWorkspace(updaterOrWorkspace, host);
-
-      return noop;
-    }
+    return result || noop;
```

Comment:
Is `updater` supposed to modify the input workspace object or return a `Rule` which updates the entire tree with a new workspace? In the first it seems like `result` is useless while the second's `writeWorkspace()` doesn't do anything?

Maybe I just don't know enough about schematics to understand why this makes sense?

Reply:
The workspace object internally tracks changes to itself and when written fine-grained updates to the underlying file are made.  This allows for comments, formatting, etc. to be maintained where possible.

The updater function can optionally return a schematic rule (or chain of rules) that will then be executed by the runtime after the updateWorkspace rule completes.  For example, this can be useful when you would want to conditionally perform some additional actions (add a dependency?) based on existing and/or changed workspace content.
I'll add that to the doc comment.

Reply:
Those sound like two different actions for `updater()`, does it need to do both? Couldn't users just put any other rules afterwards like:

```typescript
chain([
  updateWorkspace((wksp) => wksp.extensions.push({ /* ... */ })),
  addDependency('foo'),
])
```

Reply:
If adding the dependency doesn't depend on the content of the workspace, then yes.
Here's an example from one of the old migrations that leverages the functionality:
https://github.com/angular/angular-cli/blob/6723a3c91efc52a48cdc22a1681928dbf7cd1944/packages/schematics/angular/migrations/update-9/ivy-libraries.ts#L23-L60

Reply:
I see, I feel like we could use `readWorkspace()` before the `chain()` to some extent, but I could see some use cases that might be awkward with that. I don't want to hold this up, but something to consider in the future.